src/Storages/System/StorageSystemDetachedParts.cpp

```diff
@@ -8,108 +8,238 @@
         {
-            columns_mask[i] = 1;
-            header.insert(block.getByPosition(i));
+            if (detached_parts.empty())
+                getMoreParts();
+
+            if (detached_parts.empty())
+                return nullWhenNoRows(std::move(new_columns));
+
+            generateRows(new_columns, block_size - new_columns[0]->size());
+        }
+
+        return nullWhenNoRows(std::move(new_columns));
+    }
+
+private:
+    std::shared_ptr<SourceState> state;
+    const std::vector<UInt8> columns_mask;
+    const UInt64 block_size;
+    const bool has_bytes_on_disk_column;
+    const size_t support_threads = 35;
+
+    StoragesInfo current_info;
+    DetachedPartsInfo detached_parts;
+
+    void getMoreParts()
+    {
+        chassert(detached_parts.empty());
+
+        while (detached_parts.empty())
+        {
+            current_info = state->next();
+            if (!current_info)
+                return;
+
+            detached_parts = current_info.data->getDetachedParts();
+        }
+    }
+
+    void calculatePartSizeOnDisk(size_t begin, std::vector<std::atomic<size_t>> & parts_sizes)
+    {
+        if (!has_bytes_on_disk_column)
+            return;
+
+        WorkerState worker_state;
+
+        for (auto p_id = begin; p_id < detached_parts.size(); ++p_id)
+        {
+            auto & p = detached_parts.at(p_id);
+            auto & disk = p.disk;
+
+            auto part_path = fs::path(MergeTreeData::DETACHED_DIR_NAME) / p.dir_name;
+            auto relative_path = fs::path(current_info.data->getRelativeDataPath()) / part_path;
+            auto listing = partFilesOnDisk(disk, relative_path);
+
+            auto * counter = &parts_sizes[p_id - begin];
+
+            for (auto & file : listing)
+                worker_state.tasks.push_back({disk, file, counter});
+        }
+
+        std::vector<std::future<void>> futures;
+        SCOPE_EXIT_SAFE({
+            /// Cancel all workers
+            worker_state.next_task.store(worker_state.tasks.size());
+            /// Exceptions are not propagated
+            for (auto & future : futures)
+                if (future.valid())
+                    future.wait();
+            futures.clear();
+        });
+
+        auto max_thread_to_run = std::min(support_threads, worker_state.tasks.size() / 10);
+        for (size_t i = 0; i < max_thread_to_run; ++i)
+        {
+            if (worker_state.next_task.load() >= worker_state.tasks.size())
+                break;
+
+            auto worker = [&worker_state] ()
+            {
+                for (auto id = worker_state.next_task++; id < worker_state.tasks.size(); id = worker_state.next_task++)
+                {
+                    auto & task = worker_state.tasks.at(id);
+                    auto size = task.disk->getFileSize(task.file);
+                    task.counter->fetch_add(size);
+                }
+            };
+
+            futures.push_back(
```

Comment:
_Nerd mode on:_ `futures.push_back` can throw alloc exception. In this case we will not wait for the future and `WorkerState` could be destroyed before the last task ends.

Reply:
Thanks! I will just reserve the space in advance.